File: test/org/cerion/symcalc/expression/function/statistics/VarianceTest.java
package org.cerion.symcalc.expression.function.statistics;

import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.ListExpr;
import org.cerion.symcalc.expression.number.IntegerNum;
import org.cerion.symcalc.expression.number.RealNum;
import org.junit.Test;

import static org.junit.Assert.*;

public class VarianceTest {

    @Test
    public void basicIntegerTest() {
        ListExpr list = new ListExpr(new IntegerNum(600), new IntegerNum(470),
                new IntegerNum(170), new IntegerNum(430), new IntegerNum(300));

        Expr e = new Variance(list).eval();
        assertEquals(new IntegerNum(21704), e);
    }

    @Test
    public void basicTest() {
        ListExpr list = new ListExpr(RealNum.Companion.create(1.1), RealNum.Companion.create(1.2), RealNum.Companion.create(1.3));
        Expr e = new Variance(list).eval();
        assertEquals(RealNum.Companion.create(0.00666).toDouble(), ((RealNum)e).toDouble(), 0.00001);

        list = new ListExpr(RealNum.Companion.create(1.21), RealNum.Companion.create(3.4), RealNum.Companion.create(2.0),
                RealNum.Companion.create(4.66), RealNum.Companion.create(1.5), RealNum.Companion.create(5.61), RealNum.Companion.create(7.22));
        e = new Variance(list).eval();
        assertEquals(RealNum.Companion.create(4.42390).toDouble(), ((RealNum)e).toDouble(), 0.00001);
    }
}

File: test/expression/function/statistics/VarianceTest.kt
package org.cerion.symcalc.expression.function.statistics

import org.cerion.symcalc.expression.ListExpr
import org.cerion.symcalc.expression.number.Integer
import org.cerion.symcalc.expression.number.Rational
import org.cerion.symcalc.expression.number.RealDouble
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals

class VarianceTest {

    @Test
    fun basic() {
        val list = ListExpr(1.21, 3.4, 2.0, 4.66, 1.5, 5.61, 7.22)
        assertEquals(RealDouble(5.16122380952381), Variance(list).eval())
    }

    @Test
    fun integer() {
        val list = ListExpr(600, 470, 170, 430, 300)
        assertEquals(Integer(27130), Variance(list).eval())
    }

    @Test
    fun real() {
        var list = ListExpr(1.1, 1.2, 1.3)
        assertEquals(RealDouble(0.009999999999999995), Variance(list).eval())

        list = ListExpr(1.21, 3.4, 2.0, 4.66, 1.5, 5.61, 7.22)
        assertEquals(RealDouble(5.16122380952381), Variance(list).eval())
    }

    @Test
    fun integers_evalAsRational() {
        val list = ListExpr(9, 2, 5, 4, 12, 7, 8, 11, 9, 3, 7, 4, 12, 5, 4, 10, 9, 6, 9, 4)
        assertEquals(Rational(178,19), Variance(list).eval())
    }

}